### PR TITLE
update check for "all" resources

### PR DESCRIPTION
**Pull request by juanvallejo**
This patch updates the check for `all` resources to handle cases where a resource's name is "all". Rather than cycling through all given args until `all` is found, this patch makes sure that only a single argument `all` was specified at all.

**Release note**:
```release-note
release-note-none
```

@fabianofranz @pwittrock

### CODE REVIEW

File: pkg/kubectl/resource/builder.go
Comment (pwittrock) [inline pkg/kubectl/resource/builder.go]: Is args the list of args to the sub command?  Or the full list of args?

File: pkg/kubectl/resource/builder.go
Comment (janetkuo) [inline pkg/kubectl/resource/builder.go]: This check is somewhat misleading. In resource/builder.go, the call to `ReplaceAliases` replaces `all` with all supported resources [1], and then `ResourceTypeOrNameArgs` will split it into a slice of string [2]

1.`pods,replicationcontrollers,services,statefulsets,horizontalpodautoscalers,jobs,deployments,replicasets`
2.`[]string{"pods,replicationcontrollers,services,statefulsets,horizontalpodautoscalers,jobs,deployments,replicasets"}`

This means that `all` won't be passed in as `args[0]`.

The unit test needs to be updated as well.

**Comment by janetkuo** [comment]
@k8s-bot ok to test

File: pkg/kubectl/resource/builder.go
Comment (juanvallejo) [inline pkg/kubectl/resource/builder.go]: They're only the set passed to the sub command (in `kubectl get all`, `[all]` would be the args slice)

File: pkg/kubectl/resource/builder.go
Comment (juanvallejo) [inline pkg/kubectl/resource/builder.go]: > This check is somewhat misleading. In resource/builder.go, the call to ReplaceAliases replaces all with all supported resources [1], and then ResourceTypeOrNameArgs will split it into a slice of string [2]
> This means that all won't be passed in as args[0].

Although `all` will never be `args[0]` if the args slice that is passed to `MultipleTypesRequested` is built from the output of `replaceAliases` or `SplitResourceArgument`, this check is mostly for cases such as https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get.go#L424, where arguments from a sub command are passed to it, without first going through other functions in resource/builder.go

Comment (janetkuo) [inline pkg/kubectl/resource/builder.go]: > this check is mostly for cases such as https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/get.go#L424

Did you mean this line? (line number was changed since the link points to `master`):

```go
if resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, sorter) {
```

IIUC, before this line, we make a call to `ResourceTypeOrNameArgs`:

```go 
	r := resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.UnstructuredClientForMapping), unstructured.UnstructuredJSONScheme).
		NamespaceParam(cmdNamespace).DefaultNamespace().AllNamespaces(allNamespaces).
		FilenameParam(enforceNamespace, &options.FilenameOptions).
		SelectorParam(selector).
		ExportParam(export).
		ResourceTypeOrNameArgs(true, args...).
		ContinueOnError().
		Latest().
		Flatten().
		Do()
```

After this call, `args` has been changed (aliases, including `all`, are expanded). Would you explain more on when the following situation would happen? 
> arguments from a sub command are passed to it, without first going through other functions in resource/builder.go

Comment (juanvallejo) [inline pkg/kubectl/resource/builder.go]: @janetkuo 
> After this call, args has been changed (aliases, including all, are expanded).

Apologies, I see what you mean now. I had tested the values of `args` in a line located above the call to `ResourceTypeOrNameArgs` in [this PR](https://github.com/kubernetes/kubernetes/pull/39042#discussion_r96523358), and had not realized that its value is mutated by the time it is used in 
```go
if resource.MultipleTypesRequested(args) || cmdutil.MustPrintWithKinds(objs, infos, sorter) {
```

> Would you explain more on when the following situation would happen?
>>arguments from a sub command are passed to it, without first going through other functions in resource/builder.go

I realize now that, at least for the time being, `MultipleTypesRequested` will never receive `all` as one of its arguments, however a potential situation where it will is if a call to this helper is ever made before a call to `ReplaceAliases`.

Comment (janetkuo) [inline pkg/kubectl/resource/builder.go]: I see. LGTM then

**Review by janetkuo — APPROVED** [review]

**Comment by janetkuo** [comment]
Hm we need to update https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/OWNERS since @sig-cli and @sig-cli-maintainers are both removed

**Comment by janetkuo** [comment]
@juanvallejo would you squash commits? I'll apply lgtm afterwards

**Comment by juanvallejo** [comment]
@janetkuo thanks, done!

**Comment by janetkuo** [comment]
`Submit Queue — The PR was changed after the approved label was added.` 

/approve cancel

**Pull request merged**
